modular_robot/revolve2/modular_robot/body/base/_body.py

```diff
@@ -107,6 +87,66 @@ def to_grid(self) -> tuple[list[list[list[Module | None]]], tuple[int, int, int]
 
         The grid is indexed depth, width, height, or x, y, z, from the perspective of the core.
 
-        :returns: The created grid with cells set to either a Module or None and a tuple representing the position of the core.
-        :raises NotImplementedError: In case a module is encountered that is not supported.
+        :returns: The created grid with cells set to either a Module or None and a position vector of the core. The position Vector3 is dtype: int.
+        """
+        return _GridMaker().make_grid(self)
+
+    @property
+    def core(self) -> Core:
+        """
+        Get the core of the Body.
+
+        :return: The core.
         """
+        return self._core
+
+
+class _GridMaker(Generic[TModuleNP]):
+    _x: list[int] = []
+    _y: list[int] = []
+    _z: list[int] = []
+    _modules: list[Module] = []
+
+    def make_grid(self, body: Body) -> tuple[NDArray[TModuleNP], Vector3[np.int_]]:
+
+        self._make_grid_recur(body._core, Vector3(), Quaternion())
+
+        minx, maxx = min(self._x), max(self._x)
+        miny, maxy = min(self._y), max(self._y)
+        minz, maxz = min(self._z), max(self._z)
+
+        depth = maxx - minx + 1
+        width = maxy - miny + 1
+        height = maxz - minz + 1
+
+        grid = np.empty(shape=(depth, width, height), dtype=Module)
+        grid.fill(None)
+        for x, y, z, module in zip(self._x, self._y, self._z, self._modules):
+            grid[x - minx, y - miny, z - minz] = module
+
+        return grid, Vector3([-minx, -miny, -minz])
+
+    def _make_grid_recur(
+        self, module: Module, position: Vector3, orientation: Quaternion
+    ) -> None:
+        self._add(position, module)
+
+        for child_index, attachment_point in module.attachment_points.items():
+            child = module.children.get(child_index)
+            if child is not None:
+                assert np.isclose(child.rotation % (math.pi / 2.0), 0.0)
+                rotation = (
+                    orientation
+                    * attachment_point.orientation
+                    * Quaternion.from_eulers([child.rotation, 0, 0])
+                )
+                self._make_grid_recur(
+                    child, position + rotation * Vector3([1.0, 0.0, 0.0]), rotation
+                )
+
+    def _add(self, position: Vector3, module: Module) -> None:
+        self._modules.append(module)
+        x, y, z = position
+        self._x.append(round(x))
```

Comment:
I think round returns a float. your list is int. so maybe you need to do int(round(x))?